source/docs/guide/src/nonlinear_bitvec.md

```diff
@@ -0,0 +1,287 @@
+sub-classes of nonlinear formulas can be.  For example, nonlinear formulas that
+consist of a series of congruence relations (i.e., equalities modulo some
+divisor `n`).  As a simple example, we might like to show that `a % n == b % n
+==> (a * c) % n == (b * c) % n`.
+
+Verus offers a deterministic proof strategy to discharge such obligations.
+As shown below, to use this strategy, you must state the desired property
+as a proof function annotated with `by(integer_ring)`.
+
+
+Verus will then discharge the proof obligation using a dedicated algebra solver
+called [Singular](https://www.singular.uni-kl.de/).  As hinted at by the
+annotation, this proof technique is only complete (i.e., guaranteed to succeed)
+for properties that are true for all
+[rings](https://en.wikipedia.org/wiki/Ring_(mathematics)).   Formulas that rely
+specifically on properties of the integers may not be solved successfully.
+
+Using this proof technique requires a bit of additional configuration of your Verus installation.
+
+### Setup
+
+1. Install Singular
+    - To use Singular's standard library, you need more than just the Singular executable binary. 
+      Hence, when possible, we strongly recommend using your system's package manager.  Here are 
+      some suggested steps for different platforms.
+        - Mac: `brew install Singular` and set the `VERUS_SINGULAR_PATH` environment variable when running Verus. (e.g. `VERUS_SINGULAR_PATH=/usr/local/bin/Singular`). For more options, see Singular's [OS X installation guide](https://www.singular.uni-kl.de/index.php/singular-download/install-os-x.html). 
+
+        - Debian-based Linux: `apt-get install singular` and set the `VERUS_SINGULAR_PATH` environment variable when running Verus. (e.g. `VERUS_SINGULAR_PATH=/usr/bin/Singular`). For more options, see Singular's [Linux installation guide](https://www.singular.uni-kl.de/index.php/singular-download/install-linuxunix.html).
+
+        - Windows: See Singular's [Windows installation guide](https://www.singular.uni-kl.de/index.php/singular-download/install-windows.html).
+
+2. Compiling Verus with Singular Support
+    - The `integer_ring` functionality is conditionally compiled when the `singular` feature is set.
+      To add this feature, add the `--features singular` flag when you invoke `vargo build` to compile Verus.
+
+
+
+
+### Details/Limitations
+- This can be used only with **int** parameters.
+- Formulas that involve inequalities are not supported.   
+- Division is not supported.
+- Function calls in the formulas are treated as uninterpreted functions.  If a function definition is important for the proof, you should unfold the definition of the function in the proof function's `requires` clause.
+- When using `integer_ring` lemma, the divisor of `%` operator must not be zero. If a divisor can be zero in the ensures clause of the `integer_ring` lemma, the fact in ensures clause will not be available in the callsite.
```

Comment:
"When using" -> "When using an"

Reply:
"the divisor of `%` operator " -> "the divisor of a modulus operator (`%`)"
" the fact in ensures " -> " the facts in the ensures "
"in the " -> "at the"

Reply:
The bit about 0 divisors seems like an unfortunate user trap.  Could we automatically add a `requires` clause to any uses of `integer_ring` that insists that all divisors in the ensures clause (and other requires clauses?) are non-zero?